# <div dir="auto">دورة تكوينية رقم 1: التنقيب في النصوص بلغة python  القرآن نموذجا </div>

## <div dir="auto"> المرحلة الأولى: الحصول على ملف أجزاء القرآن</div>

<div dir="auto">
في مقالة  <a  href="http://abdulbaqi.io/2019/01/15/quranic_roots/" >التالية</a>، يمكن استخدام  بعض أوامر `Linux` للبحث عن الكلمات الجذرية في القرآن الكريم بأكمله أو في سورة معينة من القرآن. على الرغم من أن أوامر `Linux` يمكن أن تكون فعّالة في بعض الحالات، إلا أنها ليست مصممة لتحليل البيانات. لذا، يتعين علينا اللجوء إلى لغة برمجة للعلوم البيانية مثل R أو Python.في هذه المقالة، سأستكشف قوة `Python` لمعالجة نفس الموضوع (أي الكلمات الجذرية في القرآن)، وسأوسع في حل المشكلة لتشمل استفسارات أكثر إثارة.
</div>

<div dir="auto">
لا تهدف هذه المقالة إلى أن تكون دورة تعليمية للمبتدئين لـ `python` أو `pandas`، الحزمة الخاصة بـ `python` المخصصة لتحليل البيانات والتي سأستخدمها هنا. أتوقع أن يكون لديك بعض الخبرة مع كل منهما، ولكن يمكن أن يكون لديك خبرة أقل مع `pandas` بشكل عام.
</div>

<div dir="auto">
بدون مزيد من التأخير، لنبدأ على بركة الله.
</div>

<div dir="auto">
أولاً، دعونا نبدأ ببعض الخطوات الأساسية مثل تحميل حزمة `pandas` وإعادة تسميتها لسهولة الاستخدام باسم `pd`.
</div>

In [1]:
#! pip install pandas 

In [2]:
import pandas as pd

<div dir="auto">
ثم، لقراءة الملف الذي يحتوي على المعلومات الصرفية. تحتوي مكتبة Pandas على وظيفة read_csv التي يمكنها القراءة مباشرة من عنوان URL. يرجى ملاحظة أن الملف يحتوي على بعض المعلومات حول حقوق الملكية في السطور الأولى 56، ولذا سنستخدم خيار skiprows. أيضًا، يجب مراعاة أن read_csv يفترض افتراضيًا أن المفصل بين الحقول هو الفاصلة، وإذا كان غير ذلك - كما هو الحال في هذا الملف - يجب علينا تحديد الفاصل بشكل صريح، وبالتالي خيار sep='\t'. وأخيرًا، نقوم بعرض بضعة أسطر من الجزء العلوي باستخدام وظيفة ()head.
</div>

In [3]:
url = 'http://textminingthequran.com/data/quranic-corpus-morphology-0.4.txt'
qdforiginal = pd.read_csv(url, sep='\t',skiprows=56)
qdforiginal.head()

LOCATION        FORM  TAG                                    FEATURES
0  (1:1:1:1)          bi    P                                  PREFIX|bi+
1  (1:1:1:2)        somi    N          STEM|POS:N|LEM:{som|ROOT:smw|M|GEN
2  (1:1:2:1)     {ll~ahi   PN         STEM|POS:PN|LEM:{ll~ah|ROOT:Alh|GEN
3  (1:1:3:1)          {l  DET                                  PREFIX|Al+
4  (1:1:3:2)  r~aHoma`ni  ADJ  STEM|POS:ADJ|LEM:r~aHoma`n|ROOT:rHm|MS|GEN

<div dir="auto">
من الأحسن حفظ النسخة الأولى محليًا باستخدام وظيفة `()to_csv`.
</div>

In [4]:
qdforiginal.to_csv('quran-morphology-v1.csv')

<div dir="auto">
عند النظر إلى أول عدة أسطر في الملف أعلاه، نرى أن أعمدة `LOCATION` و `FEATURES` تحتاج إلى تقسيم إضافي.
</div>

<div dir="auto">
يحتوي ملفنا على أزيد من 128,000 سطر يمكنك التحقق من ذلك باستخدام الأمر `qdforiginal.shape`. أفضل أن نأخذ عينة صغيرة من هذا الملف الكبير ونقوم بتشغيل التجارب على عملية التقسيم. عند النجاح، يمكننا بعد ذلك تشغيلها على الملف بأكمله.
</div>

In [5]:
qdforiginal.shape

(128219, 4)

## <div dir="auto">  تقسيم الأعمدة</div>

<div dir="auto">
ٍنظرًا لأننا مهتمون  بالكلمات الجذرية، نريد تحديد كل الصفوف التي تحتوي على كلمة "ROOT:" في عمود "FEATURES". يمكن القيام بذلك باستخدام أمر مثل الأمر التالي:
</div>

In [6]:
qdforiginal.FEATURES.str.contains('ROOT')[:3]

0    False
1     True
2     True
Name: FEATURES, dtype: bool

<div dir="auto">
أخذنا الثلاثة أسطر الأولى فقط من إجمالي 128,219  سطر. يعيد ذلك قيم منطقية من نوع `True` أو `False`. لذلك، يمكننا تمرير نتيجة القيم المنطقية هذه لتصفية الإطار البيانات بأكمله بواسطة:
</div>

In [7]:
qdforiginal[qdforiginal.FEATURES.str.contains('ROOT:')].head(3)

LOCATION        FORM  TAG                                    FEATURES
1  (1:1:1:2)        somi    N          STEM|POS:N|LEM:{som|ROOT:smw|M|GEN
2  (1:1:2:1)     {ll~ahi   PN         STEM|POS:PN|LEM:{ll~ah|ROOT:Alh|GEN
4  (1:1:3:2)  r~aHoma`ni  ADJ  STEM|POS:ADJ|LEM:r~aHoma`n|ROOT:rHm|MS|GEN

<div dir="auto">
للحصول على عينة عشوائية، يمكننا دائمًا استخدام طريقة `()sample` على النحو التالي. سأسمي هذه العينة `qsample`.
</div>

In [ ]:
qsample = qdforiginal[qdforiginal.FEATURES.str.contains('ROOT:')].sample(10); qsample

<div dir="auto">
الهدف هو تقسيم عمود `LOCATION` إلى أربعة أعمدة، ثم تقسيم عمود `FEATURES` إلى عدة أعمدة كعمود للجذر .
</div>

<div dir="auto">
أولاً، سنقوم بتقسيم العمود الأول `LOCATION` إلى أربعة أعمدة. يتم ذلك من خلال استخدام وظيفة `extract` للسلسلة التي يمكنها أخذ تعبير عادي. يقوم بناء `<...>P?` ضمن التعبير العادي بإنشاء أعمدة بأسماء محددة.
</div>

In [8]:
tmp1 = qdforiginal.LOCATION.str.extract(r'(?P<sura>\d*):(?P<aya>\d*):(?P<word>\d*):(?P<w_seg>\d*)'); tmp1

sura aya word w_seg
0         1   1    1     1
1         1   1    1     2
2         1   1    2     1
3         1   1    3     1
4         1   1    3     2
...     ...  ..  ...   ...
128214  114   6    2     1
128215  114   6    2     2
128216  114   6    3     1
128217  114   6    3     2
128218  114   6    3     3

[128219 rows x 4 columns]

<div dir="auto">
الآن، دعونا نستخرج الجذور من عمود `FEATURES` بنفس الطريقة.
</div>

In [9]:
tmp2 = qdforiginal.FEATURES.str.extract(r'ROOT:(?P<Root>[^|]*)')
tmp2.fillna('', inplace=True);tmp2

Root
0          
1       smw
2       Alh
3          
4       rHm
...     ...
128214     
128215  jnn
128216     
128217     
128218  nws

[128219 rows x 1 columns]

<div dir="auto">
للحصول على الكلمات بالحروف العربية سنستخدم حزمة تم تصميمها للتنقيب عن النصوص بسهولة وتسمى <a  href="https://github.com/annotation/text-fabric" >text-fabric</a>
</div>

In [10]:
#https://github.com/q-ran/quran/blob/master/programs/morph.ipynb
from tf.writing.transcription import Transcription as tr
tmp2['RootAr'] = tmp2['Root'].map(lambda x:  tr.to_arabic(str(x)) if x!= ''  else x);tmp2

Root RootAr
0                 
1       smw    سمو
2       Alh    اله
3                 
4       rHm    رحم
...     ...    ...
128214            
128215  jnn    جنن
128216            
128217            
128218  nws    نوس

[128219 rows x 2 columns]

<div dir="auto">
بالمثل، سنستخراج الأشكال الأصلية (Lemmas) أيضًا.
</div>

In [11]:
tmp3 = qdforiginal.FEATURES.str.extract(r'LEM:(?P<Lemma>[^|]*)'); 
tmp3.fillna('', inplace=True)
tmp3['LemmaAr'] = tmp3['Lemma'].map(lambda x:  tr.to_arabic(str(x)));tmp3

Lemma    LemmaAr
0                           
1            {som       ٱسْم
2          {ll~ah     ٱللَّه
3                           
4       r~aHoma`n  رَّحْمَٰن
...           ...        ...
128214                      
128215     jin~ap     جِنَّة
128216                      
128217                      
128218      n~aAs      نَّاس

[128219 rows x 2 columns]

<div dir="auto">
كما يمكننا اضافة العمود الجديد مباشرة على الشكل التالي:
</div>

In [12]:
qdforiginal['FORMAr'] = qdforiginal['FORM'].map(lambda x: tr.to_arabic(str(x)))

In [13]:
qdforiginal

LOCATION        FORM   TAG  \
0         (1:1:1:1)          bi     P   
1         (1:1:1:2)        somi     N   
2         (1:1:2:1)     {ll~ahi    PN   
3         (1:1:3:1)          {l   DET   
4         (1:1:3:2)  r~aHoma`ni   ADJ   
...             ...         ...   ...   
128214  (114:6:2:1)         {lo   DET   
128215  (114:6:2:2)     jin~api     N   
128216  (114:6:3:1)          wa  CONJ   
128217  (114:6:3:2)          {l   DET   
128218  (114:6:3:3)      n~aAsi     N   

                                          FEATURES      FORMAr  
0                                       PREFIX|bi+          بِ  
1               STEM|POS:N|LEM:{som|ROOT:smw|M|GEN        سْمِ  
2              STEM|POS:PN|LEM:{ll~ah|ROOT:Alh|GEN     ٱللَّهِ  
3                                       PREFIX|Al+          ٱل  
4       STEM|POS:ADJ|LEM:r~aHoma`n|ROOT:rHm|MS|GEN  رَّحْمَٰنِ  
...                                            ...         ...  
128214                                  PREFIX|Al+         ٱلْ  
128215        STEM|POS:N|LEM:jin~ap|ROOT:jnn|F|GEN     جِنَّةِ  
128216                              PREFIX|w:CONJ+          وَ  
128217                                  PREFIX|Al+          ٱل  
128218        STEM|POS:N|LEM:n~aAs|ROOT:nws|MP|GEN      نَّاسِ  

[128219 rows x 5 columns]

<div dir="auto">
أخيرًا في هذه المرحلة الأولى، كل ما تبقى هو دمج `qdforiginal` مع هذه الثلاثة تقسيمات `tmp1، tmp2، tmp3`، على النحو التالي. الخيار `axis=1` يعني تشغيل الدمج على الأعمدة (وليس الصفوف).
</div>

In [14]:
df_qruan=pd.concat([tmp1, qdforiginal, tmp2, tmp3], axis=1); df_qruan

sura aya word w_seg     LOCATION        FORM   TAG  \
0         1   1    1     1    (1:1:1:1)          bi     P   
1         1   1    1     2    (1:1:1:2)        somi     N   
2         1   1    2     1    (1:1:2:1)     {ll~ahi    PN   
3         1   1    3     1    (1:1:3:1)          {l   DET   
4         1   1    3     2    (1:1:3:2)  r~aHoma`ni   ADJ   
...     ...  ..  ...   ...          ...         ...   ...   
128214  114   6    2     1  (114:6:2:1)         {lo   DET   
128215  114   6    2     2  (114:6:2:2)     jin~api     N   
128216  114   6    3     1  (114:6:3:1)          wa  CONJ   
128217  114   6    3     2  (114:6:3:2)          {l   DET   
128218  114   6    3     3  (114:6:3:3)      n~aAsi     N   

                                          FEATURES      FORMAr Root RootAr  \
0                                       PREFIX|bi+          بِ               
1               STEM|POS:N|LEM:{som|ROOT:smw|M|GEN        سْمِ  smw    سمو   
2              STEM|POS:PN|LEM:{ll~ah|ROOT:Alh|GEN     ٱللَّهِ  Alh    اله   
3                                       PREFIX|Al+          ٱل               
4       STEM|POS:ADJ|LEM:r~aHoma`n|ROOT:rHm|MS|GEN  رَّحْمَٰنِ  rHm    رحم   
...                                            ...         ...  ...    ...   
128214                                  PREFIX|Al+         ٱلْ               
128215        STEM|POS:N|LEM:jin~ap|ROOT:jnn|F|GEN     جِنَّةِ  jnn    جنن   
128216                              PREFIX|w:CONJ+          وَ               
128217                                  PREFIX|Al+          ٱل               
128218        STEM|POS:N|LEM:n~aAs|ROOT:nws|MP|GEN      نَّاسِ  nws    نوس   

            Lemma    LemmaAr  
0                             
1            {som       ٱسْم  
2          {ll~ah     ٱللَّه  
3                             
4       r~aHoma`n  رَّحْمَٰن  
...           ...        ...  
128214                        
128215     jin~ap     جِنَّة  
128216                        
128217                        
128218      n~aAs      نَّاس  

[128219 rows x 13 columns]

<div dir="auto">
لحفظ النسخة يجب إضافة التشقير التالي 'utf-8-sig'. 
</div>

In [15]:
df_qruan.to_csv('quran-morphology-v2.csv', encoding='utf-8-sig')

## <div dir="auto">  المرحلة الثانية: انشاء قاعدة البيانات للقرآن</div>

In [16]:
import sqlite3 as sql

In [17]:
### --- CREATING THE DATABASE + CURSOR
conn = sql.connect('QAC_db.db')
cur = conn.cursor()

In [ ]:
#Doping QAC table if already exists
#cur.execute("DROP TABLE QAC_TABLE")

In [18]:
### --- CREATING THE TABLE
cur.execute('''CREATE TABLE IF NOT EXISTS QAC_TABLE
           (unique_id TEXT PRIMARY KEY, sura INTEGER, aya INTEGER, word INTEGER, w_seg INTEGER, LOCATION TEXT, FORM TEXT, TAG TEXT, FEATURES TEXT, ROOT TEXT, Lemma TEXT, FORMAr TEXT,ROOTAr TEXT, LemmaAr TEXT)''')


In [19]:
### --- COPYING THE DATAFRAME TO THE TABLE
df_qruan.to_sql('QAC_TABLE', conn, if_exists='replace', index=False)

128219

In [20]:
pd.read_sql('select * from QAC_TABLE', conn)

sura aya word w_seg     LOCATION        FORM   TAG  \
0         1   1    1     1    (1:1:1:1)          bi     P   
1         1   1    1     2    (1:1:1:2)        somi     N   
2         1   1    2     1    (1:1:2:1)     {ll~ahi    PN   
3         1   1    3     1    (1:1:3:1)          {l   DET   
4         1   1    3     2    (1:1:3:2)  r~aHoma`ni   ADJ   
...     ...  ..  ...   ...          ...         ...   ...   
128214  114   6    2     1  (114:6:2:1)         {lo   DET   
128215  114   6    2     2  (114:6:2:2)     jin~api     N   
128216  114   6    3     1  (114:6:3:1)          wa  CONJ   
128217  114   6    3     2  (114:6:3:2)          {l   DET   
128218  114   6    3     3  (114:6:3:3)      n~aAsi     N   

                                          FEATURES      FORMAr Root RootAr  \
0                                       PREFIX|bi+          بِ               
1               STEM|POS:N|LEM:{som|ROOT:smw|M|GEN        سْمِ  smw    سمو   
2              STEM|POS:PN|LEM:{ll~ah|ROOT:Alh|GEN     ٱللَّهِ  Alh    اله   
3                                       PREFIX|Al+          ٱل               
4       STEM|POS:ADJ|LEM:r~aHoma`n|ROOT:rHm|MS|GEN  رَّحْمَٰنِ  rHm    رحم   
...                                            ...         ...  ...    ...   
128214                                  PREFIX|Al+         ٱلْ               
128215        STEM|POS:N|LEM:jin~ap|ROOT:jnn|F|GEN     جِنَّةِ  jnn    جنن   
128216                              PREFIX|w:CONJ+          وَ               
128217                                  PREFIX|Al+          ٱل               
128218        STEM|POS:N|LEM:n~aAs|ROOT:nws|MP|GEN      نَّاسِ  nws    نوس   

            Lemma    LemmaAr  
0                             
1            {som       ٱسْم  
2          {ll~ah     ٱللَّه  
3                             
4       r~aHoma`n  رَّحْمَٰن  
...           ...        ...  
128214                        
128215     jin~ap     جِنَّة  
128216                        
128217                        
128218      n~aAs      نَّاس  

[128219 rows x 13 columns]

# Verb Features
## Aspect features
PERF 	فعل ماض 	Perfect verb<br>
IMPF 	فعل مضارع 	Imperfect verb<br>
IMPV 	فعل أمر 	Imperative verb<br>

In [21]:
seev = pd.read_sql("select * from QAC_TABLE where TAG= 'V' and FEATURES LIKE '%IMPF%' and FEATURES LIKE '%PASS%' and FEATURES LIKE '%MOOD%' and FEATURES LIKE '%(II)%'", conn)
with pd.option_context('display.max_colwidth', None):
  display(seev.head())


sura  aya word w_seg      LOCATION       FORM TAG  \
0    2   96   21     1   (2:96:21:1)  yuEam~ara   V   
1    2  105   11     1  (2:105:11:1)  yunaz~ala   V   
2    2  272   24     1  (2:272:24:1)    yuwaf~a   V   
3    3   93   16     1   (3:93:16:1)  tunaz~ala   V   
4    5   33   12     1   (5:33:12:1)  yuqat~alu   V   

                                                         FEATURES     FORMAr  \
0  STEM|POS:V|IMPF|PASS|(II)|LEM:yuEam~aru|ROOT:Emr|3MS|MOOD:SUBJ  يُعَمَّرَ   
1    STEM|POS:V|IMPF|PASS|(II)|LEM:naz~ala|ROOT:nzl|3MS|MOOD:SUBJ  يُنَزَّلَ   
2    STEM|POS:V|IMPF|PASS|(II)|LEM:waf~aY`^|ROOT:wfy|3MS|MOOD:JUS    يُوَفَّ   
3    STEM|POS:V|IMPF|PASS|(II)|LEM:naz~ala|ROOT:nzl|3FS|MOOD:SUBJ  تُنَزَّلَ   
4    STEM|POS:V|IMPF|PASS|(II)|LEM:qut~ilu|ROOT:qtl|3MP|MOOD:SUBJ  يُقَتَّلُ   

  Root RootAr      Lemma    LemmaAr  
0  Emr    عمر  yuEam~aru  يُعَمَّرُ  
1  nzl    نزل    naz~ala    نَزَّلَ  
2  wfy    وفي   waf~aY`^   وَفَّىٰٓ  
3  nzl    نزل    naz~ala    نَزَّلَ  
4  qtl    قتل    qut~ilu    قُتِّلُ

In [22]:
allv=pd.read_sql("select * from QAC_TABLE where TAG ='V'", conn)
with pd.option_context('display.max_colwidth', None):
  display(allv.head())

sura aya word w_seg   LOCATION         FORM TAG  \
0    1   5    2     1  (1:5:2:1)     naEobudu   V   
1    1   5    4     1  (1:5:4:1)  nasotaEiynu   V   
2    1   6    1     1  (1:6:1:1)        {hodi   V   
3    1   7    3     1  (1:7:3:1)     >anoEamo   V   
4    2   3    2     1  (2:3:2:1)     yu&ominu   V   

                                         FEATURES       FORMAr Root RootAr  \
0          STEM|POS:V|IMPF|LEM:Eabada|ROOT:Ebd|1P     نَعْبُدُ  Ebd    عبد   
1  STEM|POS:V|IMPF|(X)|LEM:{sotaEiynu|ROOT:Ewn|1P  نَسْتَعِينُ  Ewn    عون   
2          STEM|POS:V|IMPV|LEM:hadaY|ROOT:hdy|2MS        ٱهْدِ  hdy    هدي   
3  STEM|POS:V|PERF|(IV)|LEM:>anoEama|ROOT:nEm|2MS     أَنْعَمْ  nEm    نعم   
4   STEM|POS:V|IMPF|(IV)|LEM:'aAmana|ROOT:Amn|3MP     يُؤْمِنُ  Amn    امن   

        Lemma     LemmaAr  
0      Eabada      عَبَدَ  
1  {sotaEiynu  ٱسْتَعِينُ  
2       hadaY       هَدَى  
3    >anoEama    أَنْعَمَ  
4     'aAmana     ءَامَنَ

In [23]:
allv['FEATURES2']=[x.split('STEM|POS:V|')[1] for x in allv['FEATURES']]
with pd.option_context('display.max_colwidth', None):
  display(allv.head())

sura aya word w_seg   LOCATION         FORM TAG  \
0    1   5    2     1  (1:5:2:1)     naEobudu   V   
1    1   5    4     1  (1:5:4:1)  nasotaEiynu   V   
2    1   6    1     1  (1:6:1:1)        {hodi   V   
3    1   7    3     1  (1:7:3:1)     >anoEamo   V   
4    2   3    2     1  (2:3:2:1)     yu&ominu   V   

                                         FEATURES       FORMAr Root RootAr  \
0          STEM|POS:V|IMPF|LEM:Eabada|ROOT:Ebd|1P     نَعْبُدُ  Ebd    عبد   
1  STEM|POS:V|IMPF|(X)|LEM:{sotaEiynu|ROOT:Ewn|1P  نَسْتَعِينُ  Ewn    عون   
2          STEM|POS:V|IMPV|LEM:hadaY|ROOT:hdy|2MS        ٱهْدِ  hdy    هدي   
3  STEM|POS:V|PERF|(IV)|LEM:>anoEama|ROOT:nEm|2MS     أَنْعَمْ  nEm    نعم   
4   STEM|POS:V|IMPF|(IV)|LEM:'aAmana|ROOT:Amn|3MP     يُؤْمِنُ  Amn    امن   

        Lemma     LemmaAr                            FEATURES2  
0      Eabada      عَبَدَ          IMPF|LEM:Eabada|ROOT:Ebd|1P  
1  {sotaEiynu  ٱسْتَعِينُ  IMPF|(X)|LEM:{sotaEiynu|ROOT:Ewn|1P  
2       hadaY       هَدَى          IMPV|LEM:hadaY|ROOT:hdy|2MS  
3    >anoEama    أَنْعَمَ  PERF|(IV)|LEM:>anoEama|ROOT:nEm|2MS  
4     'aAmana     ءَامَنَ   IMPF|(IV)|LEM:'aAmana|ROOT:Amn|3MP

In [24]:
allv['Aspect']=[x.split('|')[0] for x in allv['FEATURES2']]
with pd.option_context('display.max_colwidth', None):
  display(allv.head())

sura aya word w_seg   LOCATION         FORM TAG  \
0    1   5    2     1  (1:5:2:1)     naEobudu   V   
1    1   5    4     1  (1:5:4:1)  nasotaEiynu   V   
2    1   6    1     1  (1:6:1:1)        {hodi   V   
3    1   7    3     1  (1:7:3:1)     >anoEamo   V   
4    2   3    2     1  (2:3:2:1)     yu&ominu   V   

                                         FEATURES       FORMAr Root RootAr  \
0          STEM|POS:V|IMPF|LEM:Eabada|ROOT:Ebd|1P     نَعْبُدُ  Ebd    عبد   
1  STEM|POS:V|IMPF|(X)|LEM:{sotaEiynu|ROOT:Ewn|1P  نَسْتَعِينُ  Ewn    عون   
2          STEM|POS:V|IMPV|LEM:hadaY|ROOT:hdy|2MS        ٱهْدِ  hdy    هدي   
3  STEM|POS:V|PERF|(IV)|LEM:>anoEama|ROOT:nEm|2MS     أَنْعَمْ  nEm    نعم   
4   STEM|POS:V|IMPF|(IV)|LEM:'aAmana|ROOT:Amn|3MP     يُؤْمِنُ  Amn    امن   

        Lemma     LemmaAr                            FEATURES2 Aspect  
0      Eabada      عَبَدَ          IMPF|LEM:Eabada|ROOT:Ebd|1P   IMPF  
1  {sotaEiynu  ٱسْتَعِينُ  IMPF|(X)|LEM:{sotaEiynu|ROOT:Ewn|1P   IMPF  
2       hadaY       هَدَى          IMPV|LEM:hadaY|ROOT:hdy|2MS   IMPV  
3    >anoEama    أَنْعَمَ  PERF|(IV)|LEM:>anoEama|ROOT:nEm|2MS   PERF  
4     'aAmana     ءَامَنَ   IMPF|(IV)|LEM:'aAmana|ROOT:Amn|3MP   IMPF

##  Voice features
ACT 	مبني للمعلوم 	Active voice (default)<br>
PASS 	مبني للمجهول 	Passive voice<br>

In [25]:
allv['Voice']=[x.split('|')[1] if 'PASS' in x else 'ACT' for x in allv['FEATURES2']]; allv

sura aya word w_seg     LOCATION         FORM TAG  \
0        1   5    2     1    (1:5:2:1)     naEobudu   V   
1        1   5    4     1    (1:5:4:1)  nasotaEiynu   V   
2        1   6    1     1    (1:6:1:1)        {hodi   V   
3        1   7    3     1    (1:7:3:1)     >anoEamo   V   
4        2   3    2     1    (2:3:2:1)     yu&ominu   V   
...    ...  ..  ...   ...          ...          ...  ..   
19351  113   3    5     1  (113:3:5:1)       waqaba   V   
19352  113   5    5     1  (113:5:5:1)       Hasada   V   
19353  114   1    1     1  (114:1:1:1)         qulo   V   
19354  114   1    2     1  (114:1:2:1)      >aEuw*u   V   
19355  114   5    2     1  (114:5:2:1)   yuwasowisu   V   

                                             FEATURES       FORMAr  Root  \
0              STEM|POS:V|IMPF|LEM:Eabada|ROOT:Ebd|1P     نَعْبُدُ   Ebd   
1      STEM|POS:V|IMPF|(X)|LEM:{sotaEiynu|ROOT:Ewn|1P  نَسْتَعِينُ   Ewn   
2              STEM|POS:V|IMPV|LEM:hadaY|ROOT:hdy|2MS        ٱهْدِ   hdy   
3      STEM|POS:V|PERF|(IV)|LEM:>anoEama|ROOT:nEm|2MS     أَنْعَمْ   nEm   
4       STEM|POS:V|IMPF|(IV)|LEM:'aAmana|ROOT:Amn|3MP     يُؤْمِنُ   Amn   
...                                               ...          ...   ...   
19351         STEM|POS:V|PERF|LEM:waqaba|ROOT:wqb|3MS       وَقَبَ   wqb   
19352         STEM|POS:V|PERF|LEM:Hasada|ROOT:Hsd|3MS       حَسَدَ   Hsd   
19353          STEM|POS:V|IMPV|LEM:qaAla|ROOT:qwl|2MS         قُلْ   qwl   
19354            STEM|POS:V|IMPF|LEM:Eu*o|ROOT:Ew*|1S      أَعُوذُ   Ew*   
19355      STEM|POS:V|IMPF|LEM:wasowasa|ROOT:wsws|3MS   يُوَسْوِسُ  wsws   

      RootAr       Lemma     LemmaAr                            FEATURES2  \
0        عبد      Eabada      عَبَدَ          IMPF|LEM:Eabada|ROOT:Ebd|1P   
1        عون  {sotaEiynu  ٱسْتَعِينُ  IMPF|(X)|LEM:{sotaEiynu|ROOT:Ewn|1P   
2        هدي       hadaY       هَدَى          IMPV|LEM:hadaY|ROOT:hdy|2MS   
3        نعم    >anoEama    أَنْعَمَ  PERF|(IV)|LEM:>anoEama|ROOT:nEm|2MS   
4        امن     'aAmana     ءَامَنَ   IMPF|(IV)|LEM:'aAmana|ROOT:Amn|3MP   
...      ...         ...         ...                                  ...   
19351    وقب      waqaba      وَقَبَ         PERF|LEM:waqaba|ROOT:wqb|3MS   
19352    حسد      Hasada      حَسَدَ         PERF|LEM:Hasada|ROOT:Hsd|3MS   
19353    قول       qaAla       قَالَ          IMPV|LEM:qaAla|ROOT:qwl|2MS   
19354    عوذ        Eu*o        عُذْ            IMPF|LEM:Eu*o|ROOT:Ew*|1S   
19355   وسوس    wasowasa    وَسْوَسَ      IMPF|LEM:wasowasa|ROOT:wsws|3MS   

      Aspect Voice  
0       IMPF   ACT  
1       IMPF   ACT  
2       IMPV   ACT  
3       PERF   ACT  
4       IMPF   ACT  
...      ...   ...  
19351   PERF   ACT  
19352   PERF   ACT  
19353   IMPV   ACT  
19354   IMPF   ACT  
19355   IMPF   ACT  

[19356 rows x 16 columns]

## Mood features
IND 	مرفوع 	Indicative mood (default)<br>
SUBJ 	منصوب 	Subjunctive mood<br>
JUS 	مجزوم 	Jussive mood<br>

In [26]:
allv['MOOD']=allv['FEATURES'].map(lambda x:  x.split('|MOOD:')[1] if '|MOOD:' in x  else 'IND'); allv

sura aya word w_seg     LOCATION         FORM TAG  \
0        1   5    2     1    (1:5:2:1)     naEobudu   V   
1        1   5    4     1    (1:5:4:1)  nasotaEiynu   V   
2        1   6    1     1    (1:6:1:1)        {hodi   V   
3        1   7    3     1    (1:7:3:1)     >anoEamo   V   
4        2   3    2     1    (2:3:2:1)     yu&ominu   V   
...    ...  ..  ...   ...          ...          ...  ..   
19351  113   3    5     1  (113:3:5:1)       waqaba   V   
19352  113   5    5     1  (113:5:5:1)       Hasada   V   
19353  114   1    1     1  (114:1:1:1)         qulo   V   
19354  114   1    2     1  (114:1:2:1)      >aEuw*u   V   
19355  114   5    2     1  (114:5:2:1)   yuwasowisu   V   

                                             FEATURES       FORMAr  Root  \
0              STEM|POS:V|IMPF|LEM:Eabada|ROOT:Ebd|1P     نَعْبُدُ   Ebd   
1      STEM|POS:V|IMPF|(X)|LEM:{sotaEiynu|ROOT:Ewn|1P  نَسْتَعِينُ   Ewn   
2              STEM|POS:V|IMPV|LEM:hadaY|ROOT:hdy|2MS        ٱهْدِ   hdy   
3      STEM|POS:V|PERF|(IV)|LEM:>anoEama|ROOT:nEm|2MS     أَنْعَمْ   nEm   
4       STEM|POS:V|IMPF|(IV)|LEM:'aAmana|ROOT:Amn|3MP     يُؤْمِنُ   Amn   
...                                               ...          ...   ...   
19351         STEM|POS:V|PERF|LEM:waqaba|ROOT:wqb|3MS       وَقَبَ   wqb   
19352         STEM|POS:V|PERF|LEM:Hasada|ROOT:Hsd|3MS       حَسَدَ   Hsd   
19353          STEM|POS:V|IMPV|LEM:qaAla|ROOT:qwl|2MS         قُلْ   qwl   
19354            STEM|POS:V|IMPF|LEM:Eu*o|ROOT:Ew*|1S      أَعُوذُ   Ew*   
19355      STEM|POS:V|IMPF|LEM:wasowasa|ROOT:wsws|3MS   يُوَسْوِسُ  wsws   

      RootAr       Lemma     LemmaAr                            FEATURES2  \
0        عبد      Eabada      عَبَدَ          IMPF|LEM:Eabada|ROOT:Ebd|1P   
1        عون  {sotaEiynu  ٱسْتَعِينُ  IMPF|(X)|LEM:{sotaEiynu|ROOT:Ewn|1P   
2        هدي       hadaY       هَدَى          IMPV|LEM:hadaY|ROOT:hdy|2MS   
3        نعم    >anoEama    أَنْعَمَ  PERF|(IV)|LEM:>anoEama|ROOT:nEm|2MS   
4        امن     'aAmana     ءَامَنَ   IMPF|(IV)|LEM:'aAmana|ROOT:Amn|3MP   
...      ...         ...         ...                                  ...   
19351    وقب      waqaba      وَقَبَ         PERF|LEM:waqaba|ROOT:wqb|3MS   
19352    حسد      Hasada      حَسَدَ         PERF|LEM:Hasada|ROOT:Hsd|3MS   
19353    قول       qaAla       قَالَ          IMPV|LEM:qaAla|ROOT:qwl|2MS   
19354    عوذ        Eu*o        عُذْ            IMPF|LEM:Eu*o|ROOT:Ew*|1S   
19355   وسوس    wasowasa    وَسْوَسَ      IMPF|LEM:wasowasa|ROOT:wsws|3MS   

      Aspect Voice MOOD  
0       IMPF   ACT  IND  
1       IMPF   ACT  IND  
2       IMPV   ACT  IND  
3       PERF   ACT  IND  
4       IMPF   ACT  IND  
...      ...   ...  ...  
19351   PERF   ACT  IND  
19352   PERF   ACT  IND  
19353   IMPV   ACT  IND  
19354   IMPF   ACT  IND  
19355   IMPF   ACT  IND  

[19356 rows x 17 columns]

https://corpus.quran.com/documentation/verbforms.jsp
# Verb form features
I 	فَعَلَ First form (default)<br>
II 	فَعَّلَ Second form<br>
III 	فَاعَلَ Third form<br>
IV 	أَفْعَلَ Fourth form<br>
V 	تَفَعَّلَ Fifth form<br>
VI 	تَفَاعَلَ Sixth form<br>
VII 	إِنْفَعَلَ Seventh form<br>
VIII 	إِفْتَعَلَ Eighth form<br>
IX 	إِفْعَلَّ Ninth form<br>
X 	إِسْتَفْعَلَ Tenth form<br>
XI 	Eleventh form<br>
XII 	Twelfth form

In [27]:
import re
allv['FormV']=[re.findall(re.escape('|(')+"(.*)"+re.escape(')'),x)[0] if '|(' in x else 'I' for x in allv['FEATURES']]; allv

sura aya word w_seg     LOCATION         FORM TAG  \
0        1   5    2     1    (1:5:2:1)     naEobudu   V   
1        1   5    4     1    (1:5:4:1)  nasotaEiynu   V   
2        1   6    1     1    (1:6:1:1)        {hodi   V   
3        1   7    3     1    (1:7:3:1)     >anoEamo   V   
4        2   3    2     1    (2:3:2:1)     yu&ominu   V   
...    ...  ..  ...   ...          ...          ...  ..   
19351  113   3    5     1  (113:3:5:1)       waqaba   V   
19352  113   5    5     1  (113:5:5:1)       Hasada   V   
19353  114   1    1     1  (114:1:1:1)         qulo   V   
19354  114   1    2     1  (114:1:2:1)      >aEuw*u   V   
19355  114   5    2     1  (114:5:2:1)   yuwasowisu   V   

                                             FEATURES       FORMAr  Root  \
0              STEM|POS:V|IMPF|LEM:Eabada|ROOT:Ebd|1P     نَعْبُدُ   Ebd   
1      STEM|POS:V|IMPF|(X)|LEM:{sotaEiynu|ROOT:Ewn|1P  نَسْتَعِينُ   Ewn   
2              STEM|POS:V|IMPV|LEM:hadaY|ROOT:hdy|2MS        ٱهْدِ   hdy   
3      STEM|POS:V|PERF|(IV)|LEM:>anoEama|ROOT:nEm|2MS     أَنْعَمْ   nEm   
4       STEM|POS:V|IMPF|(IV)|LEM:'aAmana|ROOT:Amn|3MP     يُؤْمِنُ   Amn   
...                                               ...          ...   ...   
19351         STEM|POS:V|PERF|LEM:waqaba|ROOT:wqb|3MS       وَقَبَ   wqb   
19352         STEM|POS:V|PERF|LEM:Hasada|ROOT:Hsd|3MS       حَسَدَ   Hsd   
19353          STEM|POS:V|IMPV|LEM:qaAla|ROOT:qwl|2MS         قُلْ   qwl   
19354            STEM|POS:V|IMPF|LEM:Eu*o|ROOT:Ew*|1S      أَعُوذُ   Ew*   
19355      STEM|POS:V|IMPF|LEM:wasowasa|ROOT:wsws|3MS   يُوَسْوِسُ  wsws   

      RootAr       Lemma     LemmaAr                            FEATURES2  \
0        عبد      Eabada      عَبَدَ          IMPF|LEM:Eabada|ROOT:Ebd|1P   
1        عون  {sotaEiynu  ٱسْتَعِينُ  IMPF|(X)|LEM:{sotaEiynu|ROOT:Ewn|1P   
2        هدي       hadaY       هَدَى          IMPV|LEM:hadaY|ROOT:hdy|2MS   
3        نعم    >anoEama    أَنْعَمَ  PERF|(IV)|LEM:>anoEama|ROOT:nEm|2MS   
4        امن     'aAmana     ءَامَنَ   IMPF|(IV)|LEM:'aAmana|ROOT:Amn|3MP   
...      ...         ...         ...                                  ...   
19351    وقب      waqaba      وَقَبَ         PERF|LEM:waqaba|ROOT:wqb|3MS   
19352    حسد      Hasada      حَسَدَ         PERF|LEM:Hasada|ROOT:Hsd|3MS   
19353    قول       qaAla       قَالَ          IMPV|LEM:qaAla|ROOT:qwl|2MS   
19354    عوذ        Eu*o        عُذْ            IMPF|LEM:Eu*o|ROOT:Ew*|1S   
19355   وسوس    wasowasa    وَسْوَسَ      IMPF|LEM:wasowasa|ROOT:wsws|3MS   

      Aspect Voice MOOD FormV  
0       IMPF   ACT  IND     I  
1       IMPF   ACT  IND     X  
2       IMPV   ACT  IND     I  
3       PERF   ACT  IND    IV  
4       IMPF   ACT  IND    IV  
...      ...   ...  ...   ...  
19351   PERF   ACT  IND     I  
19352   PERF   ACT  IND     I  
19353   IMPV   ACT  IND     I  
19354   IMPF   ACT  IND     I  
19355   IMPF   ACT  IND     I  

[19356 rows x 18 columns]

## Person, Gender and Number

person 	الاسناد 	1, 2, 3 	first person, second person, third person<br>
gender 	الجنس 	M, F 	masculine, feminine<br>
number 	العدد 	S, D, P 	singular, dual, plural

In [28]:
allv['FEATURESR']=allv['FEATURES'].map(lambda x:  x.split('|MOOD:')[0] if '|MOOD:' in x  else x)
allv['PGN']=[x.split('|')[-1] for x in allv['FEATURESR']];allv


sura aya word w_seg     LOCATION         FORM TAG  \
0        1   5    2     1    (1:5:2:1)     naEobudu   V   
1        1   5    4     1    (1:5:4:1)  nasotaEiynu   V   
2        1   6    1     1    (1:6:1:1)        {hodi   V   
3        1   7    3     1    (1:7:3:1)     >anoEamo   V   
4        2   3    2     1    (2:3:2:1)     yu&ominu   V   
...    ...  ..  ...   ...          ...          ...  ..   
19351  113   3    5     1  (113:3:5:1)       waqaba   V   
19352  113   5    5     1  (113:5:5:1)       Hasada   V   
19353  114   1    1     1  (114:1:1:1)         qulo   V   
19354  114   1    2     1  (114:1:2:1)      >aEuw*u   V   
19355  114   5    2     1  (114:5:2:1)   yuwasowisu   V   

                                             FEATURES       FORMAr  Root  \
0              STEM|POS:V|IMPF|LEM:Eabada|ROOT:Ebd|1P     نَعْبُدُ   Ebd   
1      STEM|POS:V|IMPF|(X)|LEM:{sotaEiynu|ROOT:Ewn|1P  نَسْتَعِينُ   Ewn   
2              STEM|POS:V|IMPV|LEM:hadaY|ROOT:hdy|2MS        ٱهْدِ   hdy   
3      STEM|POS:V|PERF|(IV)|LEM:>anoEama|ROOT:nEm|2MS     أَنْعَمْ   nEm   
4       STEM|POS:V|IMPF|(IV)|LEM:'aAmana|ROOT:Amn|3MP     يُؤْمِنُ   Amn   
...                                               ...          ...   ...   
19351         STEM|POS:V|PERF|LEM:waqaba|ROOT:wqb|3MS       وَقَبَ   wqb   
19352         STEM|POS:V|PERF|LEM:Hasada|ROOT:Hsd|3MS       حَسَدَ   Hsd   
19353          STEM|POS:V|IMPV|LEM:qaAla|ROOT:qwl|2MS         قُلْ   qwl   
19354            STEM|POS:V|IMPF|LEM:Eu*o|ROOT:Ew*|1S      أَعُوذُ   Ew*   
19355      STEM|POS:V|IMPF|LEM:wasowasa|ROOT:wsws|3MS   يُوَسْوِسُ  wsws   

      RootAr       Lemma     LemmaAr                            FEATURES2  \
0        عبد      Eabada      عَبَدَ          IMPF|LEM:Eabada|ROOT:Ebd|1P   
1        عون  {sotaEiynu  ٱسْتَعِينُ  IMPF|(X)|LEM:{sotaEiynu|ROOT:Ewn|1P   
2        هدي       hadaY       هَدَى          IMPV|LEM:hadaY|ROOT:hdy|2MS   
3        نعم    >anoEama    أَنْعَمَ  PERF|(IV)|LEM:>anoEama|ROOT:nEm|2MS   
4        امن     'aAmana     ءَامَنَ   IMPF|(IV)|LEM:'aAmana|ROOT:Amn|3MP   
...      ...         ...         ...                                  ...   
19351    وقب      waqaba      وَقَبَ         PERF|LEM:waqaba|ROOT:wqb|3MS   
19352    حسد      Hasada      حَسَدَ         PERF|LEM:Hasada|ROOT:Hsd|3MS   
19353    قول       qaAla       قَالَ          IMPV|LEM:qaAla|ROOT:qwl|2MS   
19354    عوذ        Eu*o        عُذْ            IMPF|LEM:Eu*o|ROOT:Ew*|1S   
19355   وسوس    wasowasa    وَسْوَسَ      IMPF|LEM:wasowasa|ROOT:wsws|3MS   

      Aspect Voice MOOD FormV                                       FEATURESR  \
0       IMPF   ACT  IND     I          STEM|POS:V|IMPF|LEM:Eabada|ROOT:Ebd|1P   
1       IMPF   ACT  IND     X  STEM|POS:V|IMPF|(X)|LEM:{sotaEiynu|ROOT:Ewn|1P   
2       IMPV   ACT  IND     I          STEM|POS:V|IMPV|LEM:hadaY|ROOT:hdy|2MS   
3       PERF   ACT  IND    IV  STEM|POS:V|PERF|(IV)|LEM:>anoEama|ROOT:nEm|2MS   
4       IMPF   ACT  IND    IV   STEM|POS:V|IMPF|(IV)|LEM:'aAmana|ROOT:Amn|3MP   
...      ...   ...  ...   ...                                             ...   
19351   PERF   ACT  IND     I         STEM|POS:V|PERF|LEM:waqaba|ROOT:wqb|3MS   
19352   PERF   ACT  IND     I         STEM|POS:V|PERF|LEM:Hasada|ROOT:Hsd|3MS   
19353   IMPV   ACT  IND     I          STEM|POS:V|IMPV|LEM:qaAla|ROOT:qwl|2MS   
19354   IMPF   ACT  IND     I            STEM|POS:V|IMPF|LEM:Eu*o|ROOT:Ew*|1S   
19355   IMPF   ACT  IND     I      STEM|POS:V|IMPF|LEM:wasowasa|ROOT:wsws|3MS   

       PGN  
0       1P  
1       1P  
2      2MS  
3      2MS  
4      3MP  
...    ...  
19351  3MS  
19352  3MS  
19353  2MS  
19354   1S  
19355  3MS  

[19356 rows x 20 columns]

In [29]:
allv.drop(['LOCATION', 'FEATURES', 'FEATURES2', 'FEATURESR'], axis=1, inplace=True); allv

sura aya word w_seg         FORM TAG       FORMAr  Root RootAr  \
0        1   5    2     1     naEobudu   V     نَعْبُدُ   Ebd    عبد   
1        1   5    4     1  nasotaEiynu   V  نَسْتَعِينُ   Ewn    عون   
2        1   6    1     1        {hodi   V        ٱهْدِ   hdy    هدي   
3        1   7    3     1     >anoEamo   V     أَنْعَمْ   nEm    نعم   
4        2   3    2     1     yu&ominu   V     يُؤْمِنُ   Amn    امن   
...    ...  ..  ...   ...          ...  ..          ...   ...    ...   
19351  113   3    5     1       waqaba   V       وَقَبَ   wqb    وقب   
19352  113   5    5     1       Hasada   V       حَسَدَ   Hsd    حسد   
19353  114   1    1     1         qulo   V         قُلْ   qwl    قول   
19354  114   1    2     1      >aEuw*u   V      أَعُوذُ   Ew*    عوذ   
19355  114   5    2     1   yuwasowisu   V   يُوَسْوِسُ  wsws   وسوس   

            Lemma     LemmaAr Aspect Voice MOOD FormV  PGN  
0          Eabada      عَبَدَ   IMPF   ACT  IND     I   1P  
1      {sotaEiynu  ٱسْتَعِينُ   IMPF   ACT  IND     X   1P  
2           hadaY       هَدَى   IMPV   ACT  IND     I  2MS  
3        >anoEama    أَنْعَمَ   PERF   ACT  IND    IV  2MS  
4         'aAmana     ءَامَنَ   IMPF   ACT  IND    IV  3MP  
...           ...         ...    ...   ...  ...   ...  ...  
19351      waqaba      وَقَبَ   PERF   ACT  IND     I  3MS  
19352      Hasada      حَسَدَ   PERF   ACT  IND     I  3MS  
19353       qaAla       قَالَ   IMPV   ACT  IND     I  2MS  
19354        Eu*o        عُذْ   IMPF   ACT  IND     I   1S  
19355    wasowasa    وَسْوَسَ   IMPF   ACT  IND     I  3MS  

[19356 rows x 16 columns]

In [32]:
### --- CREATING THE TABLE
cur.execute('''CREATE TABLE IF NOT EXISTS QACV_TABLE
           (unique_id TEXT PRIMARY KEY, sura INTEGER, aya INTEGER, word INTEGER, w_seg INTEGER,  FORM TEXT, TAG TEXT, ROOT TEXT, Lemma TEXT, FORMAr TEXT,ROOTAr TEXT, LemmaAr TEXT, Aspect 	TEXT, MOOD TEXT, Voice TEXT, FormV TEXT, PGN TEXT)''')


In [33]:
### --- COPYING THE DATAFRAME TO THE TABLE
allv.to_sql('QACV_TABLE', conn, if_exists='replace', index=False)

19356

In [34]:
pd.read_sql('select * from QACV_TABLE', conn)

sura aya word w_seg         FORM TAG       FORMAr  Root RootAr  \
0        1   5    2     1     naEobudu   V     نَعْبُدُ   Ebd    عبد   
1        1   5    4     1  nasotaEiynu   V  نَسْتَعِينُ   Ewn    عون   
2        1   6    1     1        {hodi   V        ٱهْدِ   hdy    هدي   
3        1   7    3     1     >anoEamo   V     أَنْعَمْ   nEm    نعم   
4        2   3    2     1     yu&ominu   V     يُؤْمِنُ   Amn    امن   
...    ...  ..  ...   ...          ...  ..          ...   ...    ...   
19351  113   3    5     1       waqaba   V       وَقَبَ   wqb    وقب   
19352  113   5    5     1       Hasada   V       حَسَدَ   Hsd    حسد   
19353  114   1    1     1         qulo   V         قُلْ   qwl    قول   
19354  114   1    2     1      >aEuw*u   V      أَعُوذُ   Ew*    عوذ   
19355  114   5    2     1   yuwasowisu   V   يُوَسْوِسُ  wsws   وسوس   

            Lemma     LemmaAr Aspect Voice MOOD FormV  PGN  
0          Eabada      عَبَدَ   IMPF   ACT  IND     I   1P  
1      {sotaEiynu  ٱسْتَعِينُ   IMPF   ACT  IND     X   1P  
2           hadaY       هَدَى   IMPV   ACT  IND     I  2MS  
3        >anoEama    أَنْعَمَ   PERF   ACT  IND    IV  2MS  
4         'aAmana     ءَامَنَ   IMPF   ACT  IND    IV  3MP  
...           ...         ...    ...   ...  ...   ...  ...  
19351      waqaba      وَقَبَ   PERF   ACT  IND     I  3MS  
19352      Hasada      حَسَدَ   PERF   ACT  IND     I  3MS  
19353       qaAla       قَالَ   IMPV   ACT  IND     I  2MS  
19354        Eu*o        عُذْ   IMPF   ACT  IND     I   1S  
19355    wasowasa    وَسْوَسَ   IMPF   ACT  IND     I  3MS  

[19356 rows x 16 columns]

# Derived Nouns:
Noun (N) or adjective (ADJ) صفة part-of speech-tag with:

In [35]:
seenad=pd.read_sql("select * from QAC_TABLE where  FEATURES Like '%N|PASS%' or FEATURES Like '%ADJ%'", conn)
with pd.option_context('display.max_colwidth', None):
  display(seenad.head())

sura aya word w_seg   LOCATION         FORM  TAG  \
0    1   1    3     2  (1:1:3:2)   r~aHoma`ni  ADJ   
1    1   1    4     2  (1:1:4:2)     r~aHiymi  ADJ   
2    1   3    1     2  (1:3:1:2)   r~aHoma`ni  ADJ   
3    1   3    2     2  (1:3:2:2)     r~aHiymi  ADJ   
4    1   6    3     2  (1:6:3:2)  musotaqiyma  ADJ   

                                                   FEATURES       FORMAr Root  \
0                STEM|POS:ADJ|LEM:r~aHoma`n|ROOT:rHm|MS|GEN   رَّحْمَٰنِ  rHm   
1                  STEM|POS:ADJ|LEM:r~aHiym|ROOT:rHm|MS|GEN     رَّحِيمِ  rHm   
2                STEM|POS:ADJ|LEM:r~aHoma`n|ROOT:rHm|MS|GEN   رَّحْمَٰنِ  rHm   
3                  STEM|POS:ADJ|LEM:r~aHiym|ROOT:rHm|MS|GEN     رَّحِيمِ  rHm   
4  STEM|POS:ADJ|ACT|PCPL|(X)|LEM:m~usotaqiym|ROOT:qwm|M|ACC  مُسْتَقِيمَ  qwm   

  RootAr        Lemma      LemmaAr  
0    رحم    r~aHoma`n    رَّحْمَٰن  
1    رحم      r~aHiym      رَّحِيم  
2    رحم    r~aHoma`n    رَّحْمَٰن  
3    رحم      r~aHiym      رَّحِيم  
4    قوم  m~usotaqiym  مُّسْتَقِيم

## Derivation features
ACT PCPL 	اسم فاعل 	Active participle <br>
PASS PCPL 	اسم مفعول 	Passive participle<br>
VN 	مصدر 	Verbal noun

In [36]:
alln=pd.read_sql("select * from QAC_TABLE where TAG='N' and (FEATURES Like '%PCPL%' or FEATURES Like '%VN%') ", conn)
with pd.option_context('display.max_colwidth', None):
  display(alln.head())

sura aya word w_seg   LOCATION         FORM TAG  \
0    1   4    1     1  (1:4:1:1)      ma`liki   N   
1    1   7    6     2  (1:7:6:2)    magoDuwbi   N   
2    1   7    9     2  (1:7:9:2)  D~aA^l~iyna   N   
3    2   2    7     3  (2:2:7:3)   mut~aqiyna   N   
4    2   5    8     2  (2:5:8:2)  mufoliHuwna   N   

                                                   FEATURES       FORMAr Root  \
0             STEM|POS:N|ACT|PCPL|LEM:ma`lik|ROOT:mlk|M|GEN      مَٰلِكِ  mlk   
1          STEM|POS:N|PASS|PCPL|LEM:magoDuwb|ROOT:gDb|M|GEN    مَغْضُوبِ  gDb   
2            STEM|POS:N|ACT|PCPL|LEM:DaA^l~|ROOT:Dll|MP|GEN  ضَّآلِّينَ  Dll   
3  STEM|POS:N|ACT|PCPL|(VIII)|LEM:mut~aqiyn|ROOT:wqy|MP|GEN   مُتَّقِينَ  wqy   
4   STEM|POS:N|ACT|PCPL|(IV)|LEM:mufoliHuwn|ROOT:flH|MP|NOM  مُفْلِحُونَ  flH   

  RootAr       Lemma     LemmaAr  
0    ملك      ma`lik      مَٰلِك  
1    غضب    magoDuwb    مَغْضُوب  
2    ضلل      DaA^l~      ضَآلّ  
3    وقي   mut~aqiyn   مُتَّقِين  
4    فلح  mufoliHuwn  مُفْلِحُون

In [37]:
alladj=pd.read_sql("select * from QAC_TABLE where TAG='ADJ' and (FEATURES Like '%PCPL%' or FEATURES Like '%VN%') ", conn)

In [38]:
import re
alln['FEATURES2']=[re.findall(re.escape('STEM|POS:')+"(.*)"+re.escape('|LEM:'),x)[0] for x in alln['FEATURES']]
alln

sura aya word w_seg     LOCATION         FORM TAG  \
0       1   4    1     1    (1:4:1:1)      ma`liki   N   
1       1   7    6     2    (1:7:6:2)    magoDuwbi   N   
2       1   7    9     2    (1:7:9:2)  D~aA^l~iyna   N   
3       2   2    7     3    (2:2:7:3)   mut~aqiyna   N   
4       2   5    8     2    (2:5:8:2)  mufoliHuwna   N   
...   ...  ..  ...   ...          ...          ...  ..   
3864  109   5    3     1  (109:5:3:1)   Ea`biduwna   N   
3865  110   3    7     1  (110:3:7:1)   taw~aAbF[A   N   
3866  111   4    2     1  (111:4:2:1)   Ham~aAlapa   N   
3867  113   3    3     1  (113:3:3:1)      gaAsiqK   N   
3868  113   5    3     1  (113:5:3:1)      HaAsidK   N   

                                               FEATURES       FORMAr Root  \
0         STEM|POS:N|ACT|PCPL|LEM:ma`lik|ROOT:mlk|M|GEN      مَٰلِكِ  mlk   
1      STEM|POS:N|PASS|PCPL|LEM:magoDuwb|ROOT:gDb|M|GEN    مَغْضُوبِ  gDb   
2        STEM|POS:N|ACT|PCPL|LEM:DaA^l~|ROOT:Dll|MP|GEN  ضَّآلِّينَ  Dll   
3     STEM|POS:N|ACT|PCPL|(VIII)|LEM:mut~aqiyn|ROOT:...   مُتَّقِينَ  wqy   
4     STEM|POS:N|ACT|PCPL|(IV)|LEM:mufoliHuwn|ROOT:f...  مُفْلِحُونَ  flH   
...                                                 ...          ...  ...   
3864     STEM|POS:N|ACT|PCPL|LEM:EaAbid|ROOT:Ebd|MP|NOM   عَٰبِدُونَ  Ebd   
3865  STEM|POS:N|ACT|PCPL|LEM:taw~aAb|ROOT:twb|MS|IN...   تَوَّابًۢا  twb   
3866  STEM|POS:N|ACT|PCPL|LEM:Ham~aAlap|ROOT:Hml|FS|ACC   حَمَّالَةَ  Hml   
3867  STEM|POS:N|ACT|PCPL|LEM:gaAsiq|ROOT:gsq|M|INDE...      غَاسِقٍ  gsq   
3868  STEM|POS:N|ACT|PCPL|LEM:HaAsid|ROOT:Hsd|M|INDE...      حَاسِدٍ  Hsd   

     RootAr       Lemma     LemmaAr          FEATURES2  
0       ملك      ma`lik      مَٰلِك         N|ACT|PCPL  
1       غضب    magoDuwb    مَغْضُوب        N|PASS|PCPL  
2       ضلل      DaA^l~      ضَآلّ         N|ACT|PCPL  
3       وقي   mut~aqiyn   مُتَّقِين  N|ACT|PCPL|(VIII)  
4       فلح  mufoliHuwn  مُفْلِحُون    N|ACT|PCPL|(IV)  
...     ...         ...         ...                ...  
3864    عبد      EaAbid      عَابِد         N|ACT|PCPL  
3865    توب     taw~aAb     تَوَّاب         N|ACT|PCPL  
3866    حمل   Ham~aAlap   حَمَّالَة         N|ACT|PCPL  
3867    غسق      gaAsiq      غَاسِق         N|ACT|PCPL  
3868    حسد      HaAsid      حَاسِد         N|ACT|PCPL  

[3869 rows x 14 columns]

In [39]:
alladj['FEATURES2']=[re.findall(re.escape('STEM|POS:')+"(.*)"+re.escape('|LEM:'),x)[0] for x in alladj['FEATURES']]

In [40]:
alln['Derivation']=alln['FEATURES2'].map(lambda x:  x.split('|')[1]+' '+'PCPL' if '|PCPL' in x  else 'VN'); alln

sura aya word w_seg     LOCATION         FORM TAG  \
0       1   4    1     1    (1:4:1:1)      ma`liki   N   
1       1   7    6     2    (1:7:6:2)    magoDuwbi   N   
2       1   7    9     2    (1:7:9:2)  D~aA^l~iyna   N   
3       2   2    7     3    (2:2:7:3)   mut~aqiyna   N   
4       2   5    8     2    (2:5:8:2)  mufoliHuwna   N   
...   ...  ..  ...   ...          ...          ...  ..   
3864  109   5    3     1  (109:5:3:1)   Ea`biduwna   N   
3865  110   3    7     1  (110:3:7:1)   taw~aAbF[A   N   
3866  111   4    2     1  (111:4:2:1)   Ham~aAlapa   N   
3867  113   3    3     1  (113:3:3:1)      gaAsiqK   N   
3868  113   5    3     1  (113:5:3:1)      HaAsidK   N   

                                               FEATURES       FORMAr Root  \
0         STEM|POS:N|ACT|PCPL|LEM:ma`lik|ROOT:mlk|M|GEN      مَٰلِكِ  mlk   
1      STEM|POS:N|PASS|PCPL|LEM:magoDuwb|ROOT:gDb|M|GEN    مَغْضُوبِ  gDb   
2        STEM|POS:N|ACT|PCPL|LEM:DaA^l~|ROOT:Dll|MP|GEN  ضَّآلِّينَ  Dll   
3     STEM|POS:N|ACT|PCPL|(VIII)|LEM:mut~aqiyn|ROOT:...   مُتَّقِينَ  wqy   
4     STEM|POS:N|ACT|PCPL|(IV)|LEM:mufoliHuwn|ROOT:f...  مُفْلِحُونَ  flH   
...                                                 ...          ...  ...   
3864     STEM|POS:N|ACT|PCPL|LEM:EaAbid|ROOT:Ebd|MP|NOM   عَٰبِدُونَ  Ebd   
3865  STEM|POS:N|ACT|PCPL|LEM:taw~aAb|ROOT:twb|MS|IN...   تَوَّابًۢا  twb   
3866  STEM|POS:N|ACT|PCPL|LEM:Ham~aAlap|ROOT:Hml|FS|ACC   حَمَّالَةَ  Hml   
3867  STEM|POS:N|ACT|PCPL|LEM:gaAsiq|ROOT:gsq|M|INDE...      غَاسِقٍ  gsq   
3868  STEM|POS:N|ACT|PCPL|LEM:HaAsid|ROOT:Hsd|M|INDE...      حَاسِدٍ  Hsd   

     RootAr       Lemma     LemmaAr          FEATURES2 Derivation  
0       ملك      ma`lik      مَٰلِك         N|ACT|PCPL   ACT PCPL  
1       غضب    magoDuwb    مَغْضُوب        N|PASS|PCPL  PASS PCPL  
2       ضلل      DaA^l~      ضَآلّ         N|ACT|PCPL   ACT PCPL  
3       وقي   mut~aqiyn   مُتَّقِين  N|ACT|PCPL|(VIII)   ACT PCPL  
4       فلح  mufoliHuwn  مُفْلِحُون    N|ACT|PCPL|(IV)   ACT PCPL  
...     ...         ...         ...                ...        ...  
3864    عبد      EaAbid      عَابِد         N|ACT|PCPL   ACT PCPL  
3865    توب     taw~aAb     تَوَّاب         N|ACT|PCPL   ACT PCPL  
3866    حمل   Ham~aAlap   حَمَّالَة         N|ACT|PCPL   ACT PCPL  
3867    غسق      gaAsiq      غَاسِق         N|ACT|PCPL   ACT PCPL  
3868    حسد      HaAsid      حَاسِد         N|ACT|PCPL   ACT PCPL  

[3869 rows x 15 columns]

In [41]:
alladj['Derivation']=alladj['FEATURES2'].map(lambda x:  x.split('|')[1]+' '+'PCPL' if '|PCPL' in x  else 'VN')

# Nominal Features:
## State features
DEF 	معرفة 	Definite state; <br>
INDEF 	نكرة 	Indefinite state

In [42]:
indef=pd.read_sql("select * from QAC_TABLE where TAG ='N' and FEATURES NOT Like '%INDEF%' ", conn)
with pd.option_context('display.max_colwidth', None):
  display(indef.head())

sura aya word w_seg   LOCATION        FORM TAG  \
0    1   1    1     2  (1:1:1:2)        somi   N   
1    1   2    1     2  (1:2:1:2)      Hamodu   N   
2    1   2    3     1  (1:2:3:1)       rab~i   N   
3    1   2    4     2  (1:2:4:2)  Ea`lamiyna   N   
4    1   4    1     1  (1:4:1:1)     ma`liki   N   

                                        FEATURES      FORMAr Root RootAr  \
0             STEM|POS:N|LEM:{som|ROOT:smw|M|GEN        سْمِ  smw    سمو   
1            STEM|POS:N|LEM:Hamod|ROOT:Hmd|M|NOM      حَمْدُ  Hmd    حمد   
2             STEM|POS:N|LEM:rab~|ROOT:rbb|M|GEN       رَبِّ  rbb    ربب   
3       STEM|POS:N|LEM:Ea`lamiyn|ROOT:Elm|MP|GEN  عَٰلَمِينَ  Elm    علم   
4  STEM|POS:N|ACT|PCPL|LEM:ma`lik|ROOT:mlk|M|GEN     مَٰلِكِ  mlk    ملك   

       Lemma    LemmaAr  
0       {som       ٱسْم  
1      Hamod      حَمْد  
2       rab~       رَبّ  
3  Ea`lamiyn  عَٰلَمِين  
4     ma`lik     مَٰلِك

In [43]:
alln['State']=alln['FEATURES'].map(lambda x:  'INDEF' if '|INDEF' in x  else 'DEF'); alln

sura aya word w_seg     LOCATION         FORM TAG  \
0       1   4    1     1    (1:4:1:1)      ma`liki   N   
1       1   7    6     2    (1:7:6:2)    magoDuwbi   N   
2       1   7    9     2    (1:7:9:2)  D~aA^l~iyna   N   
3       2   2    7     3    (2:2:7:3)   mut~aqiyna   N   
4       2   5    8     2    (2:5:8:2)  mufoliHuwna   N   
...   ...  ..  ...   ...          ...          ...  ..   
3864  109   5    3     1  (109:5:3:1)   Ea`biduwna   N   
3865  110   3    7     1  (110:3:7:1)   taw~aAbF[A   N   
3866  111   4    2     1  (111:4:2:1)   Ham~aAlapa   N   
3867  113   3    3     1  (113:3:3:1)      gaAsiqK   N   
3868  113   5    3     1  (113:5:3:1)      HaAsidK   N   

                                               FEATURES       FORMAr Root  \
0         STEM|POS:N|ACT|PCPL|LEM:ma`lik|ROOT:mlk|M|GEN      مَٰلِكِ  mlk   
1      STEM|POS:N|PASS|PCPL|LEM:magoDuwb|ROOT:gDb|M|GEN    مَغْضُوبِ  gDb   
2        STEM|POS:N|ACT|PCPL|LEM:DaA^l~|ROOT:Dll|MP|GEN  ضَّآلِّينَ  Dll   
3     STEM|POS:N|ACT|PCPL|(VIII)|LEM:mut~aqiyn|ROOT:...   مُتَّقِينَ  wqy   
4     STEM|POS:N|ACT|PCPL|(IV)|LEM:mufoliHuwn|ROOT:f...  مُفْلِحُونَ  flH   
...                                                 ...          ...  ...   
3864     STEM|POS:N|ACT|PCPL|LEM:EaAbid|ROOT:Ebd|MP|NOM   عَٰبِدُونَ  Ebd   
3865  STEM|POS:N|ACT|PCPL|LEM:taw~aAb|ROOT:twb|MS|IN...   تَوَّابًۢا  twb   
3866  STEM|POS:N|ACT|PCPL|LEM:Ham~aAlap|ROOT:Hml|FS|ACC   حَمَّالَةَ  Hml   
3867  STEM|POS:N|ACT|PCPL|LEM:gaAsiq|ROOT:gsq|M|INDE...      غَاسِقٍ  gsq   
3868  STEM|POS:N|ACT|PCPL|LEM:HaAsid|ROOT:Hsd|M|INDE...      حَاسِدٍ  Hsd   

     RootAr       Lemma     LemmaAr          FEATURES2 Derivation  State  
0       ملك      ma`lik      مَٰلِك         N|ACT|PCPL   ACT PCPL    DEF  
1       غضب    magoDuwb    مَغْضُوب        N|PASS|PCPL  PASS PCPL    DEF  
2       ضلل      DaA^l~      ضَآلّ         N|ACT|PCPL   ACT PCPL    DEF  
3       وقي   mut~aqiyn   مُتَّقِين  N|ACT|PCPL|(VIII)   ACT PCPL    DEF  
4       فلح  mufoliHuwn  مُفْلِحُون    N|ACT|PCPL|(IV)   ACT PCPL    DEF  
...     ...         ...         ...                ...        ...    ...  
3864    عبد      EaAbid      عَابِد         N|ACT|PCPL   ACT PCPL    DEF  
3865    توب     taw~aAb     تَوَّاب         N|ACT|PCPL   ACT PCPL  INDEF  
3866    حمل   Ham~aAlap   حَمَّالَة         N|ACT|PCPL   ACT PCPL    DEF  
3867    غسق      gaAsiq      غَاسِق         N|ACT|PCPL   ACT PCPL  INDEF  
3868    حسد      HaAsid      حَاسِد         N|ACT|PCPL   ACT PCPL  INDEF  

[3869 rows x 16 columns]

In [44]:
alladj['State']=alladj['FEATURES'].map(lambda x:  'INDEF' if '|INDEF' in x  else 'DEF')

## Case features
NOM 	مرفوع 	Nominative case;<br>
ACC 	منصوب 	Accusative case;<br>
GEN 	مجرور 	Genitive case

extract string between two strings python

In [45]:
alln['FEATURESR']=[x.split('ROOT:')[1] for x in alln['FEATURES']]
alln['FEATURESR']=[x.split('|')[1:] for x in alln['FEATURESR']]
with pd.option_context('display.max_colwidth', None):
  display(alln.head())

sura aya word w_seg   LOCATION         FORM TAG  \
0    1   4    1     1  (1:4:1:1)      ma`liki   N   
1    1   7    6     2  (1:7:6:2)    magoDuwbi   N   
2    1   7    9     2  (1:7:9:2)  D~aA^l~iyna   N   
3    2   2    7     3  (2:2:7:3)   mut~aqiyna   N   
4    2   5    8     2  (2:5:8:2)  mufoliHuwna   N   

                                                   FEATURES       FORMAr Root  \
0             STEM|POS:N|ACT|PCPL|LEM:ma`lik|ROOT:mlk|M|GEN      مَٰلِكِ  mlk   
1          STEM|POS:N|PASS|PCPL|LEM:magoDuwb|ROOT:gDb|M|GEN    مَغْضُوبِ  gDb   
2            STEM|POS:N|ACT|PCPL|LEM:DaA^l~|ROOT:Dll|MP|GEN  ضَّآلِّينَ  Dll   
3  STEM|POS:N|ACT|PCPL|(VIII)|LEM:mut~aqiyn|ROOT:wqy|MP|GEN   مُتَّقِينَ  wqy   
4   STEM|POS:N|ACT|PCPL|(IV)|LEM:mufoliHuwn|ROOT:flH|MP|NOM  مُفْلِحُونَ  flH   

  RootAr       Lemma     LemmaAr          FEATURES2 Derivation State  \
0    ملك      ma`lik      مَٰلِك         N|ACT|PCPL   ACT PCPL   DEF   
1    غضب    magoDuwb    مَغْضُوب        N|PASS|PCPL  PASS PCPL   DEF   
2    ضلل      DaA^l~      ضَآلّ         N|ACT|PCPL   ACT PCPL   DEF   
3    وقي   mut~aqiyn   مُتَّقِين  N|ACT|PCPL|(VIII)   ACT PCPL   DEF   
4    فلح  mufoliHuwn  مُفْلِحُون    N|ACT|PCPL|(IV)   ACT PCPL   DEF   

   FEATURESR  
0   [M, GEN]  
1   [M, GEN]  
2  [MP, GEN]  
3  [MP, GEN]  
4  [MP, NOM]

In [46]:
alladj['FEATURESR']=[x.split('ROOT:')[1] for x in alladj['FEATURES']]
alladj['FEATURESR']=[x.split('|')[1:] for x in alladj['FEATURESR']]

In [47]:
crx=alladj[alladj['LOCATION'] == '(2:69:15:1)'] 
with pd.option_context('display.max_colwidth', None):
  display(crx)

sura aya word w_seg     LOCATION     FORM  TAG  \
4    2  69   15     1  (2:69:15:1)  faAqiEN  ADJ   

                                              FEATURES   FORMAr Root RootAr  \
4  STEM|POS:ADJ|ACT|PCPL|LEM:faAqiE|ROOT:fqE|INDEF|NOM  فَاقِعٌ  fqE    فقع   

    Lemma LemmaAr     FEATURES2 Derivation  State     FEATURESR  
4  faAqiE  فَاقِع  ADJ|ACT|PCPL   ACT PCPL  INDEF  [INDEF, NOM]

In [48]:
alln['Grop']=[x[0] if 'INDEF' not in str(x) and len(x)>1 else '' for x in alln['FEATURESR']]
alln['Case']=[str(x[-1]) for x in alln['FEATURESR']]
alln

sura aya word w_seg     LOCATION         FORM TAG  \
0       1   4    1     1    (1:4:1:1)      ma`liki   N   
1       1   7    6     2    (1:7:6:2)    magoDuwbi   N   
2       1   7    9     2    (1:7:9:2)  D~aA^l~iyna   N   
3       2   2    7     3    (2:2:7:3)   mut~aqiyna   N   
4       2   5    8     2    (2:5:8:2)  mufoliHuwna   N   
...   ...  ..  ...   ...          ...          ...  ..   
3864  109   5    3     1  (109:5:3:1)   Ea`biduwna   N   
3865  110   3    7     1  (110:3:7:1)   taw~aAbF[A   N   
3866  111   4    2     1  (111:4:2:1)   Ham~aAlapa   N   
3867  113   3    3     1  (113:3:3:1)      gaAsiqK   N   
3868  113   5    3     1  (113:5:3:1)      HaAsidK   N   

                                               FEATURES       FORMAr Root  \
0         STEM|POS:N|ACT|PCPL|LEM:ma`lik|ROOT:mlk|M|GEN      مَٰلِكِ  mlk   
1      STEM|POS:N|PASS|PCPL|LEM:magoDuwb|ROOT:gDb|M|GEN    مَغْضُوبِ  gDb   
2        STEM|POS:N|ACT|PCPL|LEM:DaA^l~|ROOT:Dll|MP|GEN  ضَّآلِّينَ  Dll   
3     STEM|POS:N|ACT|PCPL|(VIII)|LEM:mut~aqiyn|ROOT:...   مُتَّقِينَ  wqy   
4     STEM|POS:N|ACT|PCPL|(IV)|LEM:mufoliHuwn|ROOT:f...  مُفْلِحُونَ  flH   
...                                                 ...          ...  ...   
3864     STEM|POS:N|ACT|PCPL|LEM:EaAbid|ROOT:Ebd|MP|NOM   عَٰبِدُونَ  Ebd   
3865  STEM|POS:N|ACT|PCPL|LEM:taw~aAb|ROOT:twb|MS|IN...   تَوَّابًۢا  twb   
3866  STEM|POS:N|ACT|PCPL|LEM:Ham~aAlap|ROOT:Hml|FS|ACC   حَمَّالَةَ  Hml   
3867  STEM|POS:N|ACT|PCPL|LEM:gaAsiq|ROOT:gsq|M|INDE...      غَاسِقٍ  gsq   
3868  STEM|POS:N|ACT|PCPL|LEM:HaAsid|ROOT:Hsd|M|INDE...      حَاسِدٍ  Hsd   

     RootAr       Lemma     LemmaAr          FEATURES2 Derivation  State  \
0       ملك      ma`lik      مَٰلِك         N|ACT|PCPL   ACT PCPL    DEF   
1       غضب    magoDuwb    مَغْضُوب        N|PASS|PCPL  PASS PCPL    DEF   
2       ضلل      DaA^l~      ضَآلّ         N|ACT|PCPL   ACT PCPL    DEF   
3       وقي   mut~aqiyn   مُتَّقِين  N|ACT|PCPL|(VIII)   ACT PCPL    DEF   
4       فلح  mufoliHuwn  مُفْلِحُون    N|ACT|PCPL|(IV)   ACT PCPL    DEF   
...     ...         ...         ...                ...        ...    ...   
3864    عبد      EaAbid      عَابِد         N|ACT|PCPL   ACT PCPL    DEF   
3865    توب     taw~aAb     تَوَّاب         N|ACT|PCPL   ACT PCPL  INDEF   
3866    حمل   Ham~aAlap   حَمَّالَة         N|ACT|PCPL   ACT PCPL    DEF   
3867    غسق      gaAsiq      غَاسِق         N|ACT|PCPL   ACT PCPL  INDEF   
3868    حسد      HaAsid      حَاسِد         N|ACT|PCPL   ACT PCPL  INDEF   

             FEATURESR Grop Case  
0             [M, GEN]    M  GEN  
1             [M, GEN]    M  GEN  
2            [MP, GEN]   MP  GEN  
3            [MP, GEN]   MP  GEN  
4            [MP, NOM]   MP  NOM  
...                ...  ...  ...  
3864         [MP, NOM]   MP  NOM  
3865  [MS, INDEF, ACC]       ACC  
3866         [FS, ACC]   FS  ACC  
3867   [M, INDEF, GEN]       GEN  
3868   [M, INDEF, GEN]       GEN  

[3869 rows x 19 columns]

In [49]:
alladj['Grop']=[x[0] if 'INDEF' not in str(x) else '' for x in alladj['FEATURESR']]
alladj['Case']=[str(x[-1]) for x in alladj['FEATURESR']]

In [50]:
alln['FormN']=[re.findall(re.escape('|(')+"(.*)"+re.escape(')'),x)[0] if '|(' in x else 'I' for x in alln['FEATURES']]
alln

sura aya word w_seg     LOCATION         FORM TAG  \
0       1   4    1     1    (1:4:1:1)      ma`liki   N   
1       1   7    6     2    (1:7:6:2)    magoDuwbi   N   
2       1   7    9     2    (1:7:9:2)  D~aA^l~iyna   N   
3       2   2    7     3    (2:2:7:3)   mut~aqiyna   N   
4       2   5    8     2    (2:5:8:2)  mufoliHuwna   N   
...   ...  ..  ...   ...          ...          ...  ..   
3864  109   5    3     1  (109:5:3:1)   Ea`biduwna   N   
3865  110   3    7     1  (110:3:7:1)   taw~aAbF[A   N   
3866  111   4    2     1  (111:4:2:1)   Ham~aAlapa   N   
3867  113   3    3     1  (113:3:3:1)      gaAsiqK   N   
3868  113   5    3     1  (113:5:3:1)      HaAsidK   N   

                                               FEATURES       FORMAr Root  \
0         STEM|POS:N|ACT|PCPL|LEM:ma`lik|ROOT:mlk|M|GEN      مَٰلِكِ  mlk   
1      STEM|POS:N|PASS|PCPL|LEM:magoDuwb|ROOT:gDb|M|GEN    مَغْضُوبِ  gDb   
2        STEM|POS:N|ACT|PCPL|LEM:DaA^l~|ROOT:Dll|MP|GEN  ضَّآلِّينَ  Dll   
3     STEM|POS:N|ACT|PCPL|(VIII)|LEM:mut~aqiyn|ROOT:...   مُتَّقِينَ  wqy   
4     STEM|POS:N|ACT|PCPL|(IV)|LEM:mufoliHuwn|ROOT:f...  مُفْلِحُونَ  flH   
...                                                 ...          ...  ...   
3864     STEM|POS:N|ACT|PCPL|LEM:EaAbid|ROOT:Ebd|MP|NOM   عَٰبِدُونَ  Ebd   
3865  STEM|POS:N|ACT|PCPL|LEM:taw~aAb|ROOT:twb|MS|IN...   تَوَّابًۢا  twb   
3866  STEM|POS:N|ACT|PCPL|LEM:Ham~aAlap|ROOT:Hml|FS|ACC   حَمَّالَةَ  Hml   
3867  STEM|POS:N|ACT|PCPL|LEM:gaAsiq|ROOT:gsq|M|INDE...      غَاسِقٍ  gsq   
3868  STEM|POS:N|ACT|PCPL|LEM:HaAsid|ROOT:Hsd|M|INDE...      حَاسِدٍ  Hsd   

     RootAr       Lemma     LemmaAr          FEATURES2 Derivation  State  \
0       ملك      ma`lik      مَٰلِك         N|ACT|PCPL   ACT PCPL    DEF   
1       غضب    magoDuwb    مَغْضُوب        N|PASS|PCPL  PASS PCPL    DEF   
2       ضلل      DaA^l~      ضَآلّ         N|ACT|PCPL   ACT PCPL    DEF   
3       وقي   mut~aqiyn   مُتَّقِين  N|ACT|PCPL|(VIII)   ACT PCPL    DEF   
4       فلح  mufoliHuwn  مُفْلِحُون    N|ACT|PCPL|(IV)   ACT PCPL    DEF   
...     ...         ...         ...                ...        ...    ...   
3864    عبد      EaAbid      عَابِد         N|ACT|PCPL   ACT PCPL    DEF   
3865    توب     taw~aAb     تَوَّاب         N|ACT|PCPL   ACT PCPL  INDEF   
3866    حمل   Ham~aAlap   حَمَّالَة         N|ACT|PCPL   ACT PCPL    DEF   
3867    غسق      gaAsiq      غَاسِق         N|ACT|PCPL   ACT PCPL  INDEF   
3868    حسد      HaAsid      حَاسِد         N|ACT|PCPL   ACT PCPL  INDEF   

             FEATURESR Grop Case FormN  
0             [M, GEN]    M  GEN     I  
1             [M, GEN]    M  GEN     I  
2            [MP, GEN]   MP  GEN     I  
3            [MP, GEN]   MP  GEN  VIII  
4            [MP, NOM]   MP  NOM    IV  
...                ...  ...  ...   ...  
3864         [MP, NOM]   MP  NOM     I  
3865  [MS, INDEF, ACC]       ACC     I  
3866         [FS, ACC]   FS  ACC     I  
3867   [M, INDEF, GEN]       GEN     I  
3868   [M, INDEF, GEN]       GEN     I  

[3869 rows x 20 columns]

In [51]:
alladj['FormN']=[re.findall(re.escape('|(')+"(.*)"+re.escape(')'),x)[0] if '|(' in x else 'I' for x in alladj['FEATURES']]

In [52]:
alln.drop(['LOCATION', 'FEATURES', 'FEATURES2', 'FEATURESR'], axis=1, inplace=True); alln

sura aya word w_seg         FORM TAG       FORMAr Root RootAr  \
0       1   4    1     1      ma`liki   N      مَٰلِكِ  mlk    ملك   
1       1   7    6     2    magoDuwbi   N    مَغْضُوبِ  gDb    غضب   
2       1   7    9     2  D~aA^l~iyna   N  ضَّآلِّينَ  Dll    ضلل   
3       2   2    7     3   mut~aqiyna   N   مُتَّقِينَ  wqy    وقي   
4       2   5    8     2  mufoliHuwna   N  مُفْلِحُونَ  flH    فلح   
...   ...  ..  ...   ...          ...  ..          ...  ...    ...   
3864  109   5    3     1   Ea`biduwna   N   عَٰبِدُونَ  Ebd    عبد   
3865  110   3    7     1   taw~aAbF[A   N   تَوَّابًۢا  twb    توب   
3866  111   4    2     1   Ham~aAlapa   N   حَمَّالَةَ  Hml    حمل   
3867  113   3    3     1      gaAsiqK   N      غَاسِقٍ  gsq    غسق   
3868  113   5    3     1      HaAsidK   N      حَاسِدٍ  Hsd    حسد   

           Lemma     LemmaAr Derivation  State Grop Case FormN  
0         ma`lik      مَٰلِك   ACT PCPL    DEF    M  GEN     I  
1       magoDuwb    مَغْضُوب  PASS PCPL    DEF    M  GEN     I  
2         DaA^l~      ضَآلّ   ACT PCPL    DEF   MP  GEN     I  
3      mut~aqiyn   مُتَّقِين   ACT PCPL    DEF   MP  GEN  VIII  
4     mufoliHuwn  مُفْلِحُون   ACT PCPL    DEF   MP  NOM    IV  
...          ...         ...        ...    ...  ...  ...   ...  
3864      EaAbid      عَابِد   ACT PCPL    DEF   MP  NOM     I  
3865     taw~aAb     تَوَّاب   ACT PCPL  INDEF       ACC     I  
3866   Ham~aAlap   حَمَّالَة   ACT PCPL    DEF   FS  ACC     I  
3867      gaAsiq      غَاسِق   ACT PCPL  INDEF       GEN     I  
3868      HaAsid      حَاسِد   ACT PCPL  INDEF       GEN     I  

[3869 rows x 16 columns]

In [53]:
alladj.drop(['LOCATION', 'FEATURES', 'FEATURES2', 'FEATURESR'], axis=1, inplace=True)

In [54]:
### --- CREATING THE TABLE
cur.execute('''CREATE TABLE IF NOT EXISTS QACN_TABLE
           (unique_id TEXT PRIMARY KEY, sura INTEGER, aya INTEGER, word INTEGER, w_seg INTEGER,  FORM TEXT, TAG TEXT, ROOT TEXT, Lemma TEXT, FORMAr TEXT, ROOTAr TEXT, LemmaAr TEXT, Derivation TEXT, State TEXT, Grop TEXT, Casen TEXT, FormN TEXT )''')


In [55]:
cur.execute('''CREATE TABLE IF NOT EXISTS QACA_TABLE
           (unique_id TEXT PRIMARY KEY, sura INTEGER, aya INTEGER, word INTEGER, w_seg INTEGER,  FORM TEXT, TAG TEXT, ROOT TEXT, Lemma TEXT, FORMAr TEXT, ROOTAr TEXT, LemmaAr TEXT, Derivation TEXT, State TEXT, Grop TEXT, Casen TEXT, FormN TEXT )''')


In [56]:
### --- COPYING THE DATAFRAME TO THE TABLE
alln.to_sql('QACN_TABLE', conn, if_exists='replace', index=False)

3869

In [57]:
### --- COPYING THE DATAFRAME TO THE TABLE
alladj.to_sql('QACA_TABLE', conn, if_exists='replace', index=False)

335

In [58]:
pd.read_sql('select * from QACN_TABLE', conn)

sura aya word w_seg         FORM TAG       FORMAr Root RootAr  \
0       1   4    1     1      ma`liki   N      مَٰلِكِ  mlk    ملك   
1       1   7    6     2    magoDuwbi   N    مَغْضُوبِ  gDb    غضب   
2       1   7    9     2  D~aA^l~iyna   N  ضَّآلِّينَ  Dll    ضلل   
3       2   2    7     3   mut~aqiyna   N   مُتَّقِينَ  wqy    وقي   
4       2   5    8     2  mufoliHuwna   N  مُفْلِحُونَ  flH    فلح   
...   ...  ..  ...   ...          ...  ..          ...  ...    ...   
3864  109   5    3     1   Ea`biduwna   N   عَٰبِدُونَ  Ebd    عبد   
3865  110   3    7     1   taw~aAbF[A   N   تَوَّابًۢا  twb    توب   
3866  111   4    2     1   Ham~aAlapa   N   حَمَّالَةَ  Hml    حمل   
3867  113   3    3     1      gaAsiqK   N      غَاسِقٍ  gsq    غسق   
3868  113   5    3     1      HaAsidK   N      حَاسِدٍ  Hsd    حسد   

           Lemma     LemmaAr Derivation  State Grop Case FormN  
0         ma`lik      مَٰلِك   ACT PCPL    DEF    M  GEN     I  
1       magoDuwb    مَغْضُوب  PASS PCPL    DEF    M  GEN     I  
2         DaA^l~      ضَآلّ   ACT PCPL    DEF   MP  GEN     I  
3      mut~aqiyn   مُتَّقِين   ACT PCPL    DEF   MP  GEN  VIII  
4     mufoliHuwn  مُفْلِحُون   ACT PCPL    DEF   MP  NOM    IV  
...          ...         ...        ...    ...  ...  ...   ...  
3864      EaAbid      عَابِد   ACT PCPL    DEF   MP  NOM     I  
3865     taw~aAb     تَوَّاب   ACT PCPL  INDEF       ACC     I  
3866   Ham~aAlap   حَمَّالَة   ACT PCPL    DEF   FS  ACC     I  
3867      gaAsiq      غَاسِق   ACT PCPL  INDEF       GEN     I  
3868      HaAsid      حَاسِد   ACT PCPL  INDEF       GEN     I  

[3869 rows x 16 columns]

In [59]:
pd.read_sql('select * from QACA_TABLE', conn)

sura aya word w_seg           FORM  TAG         FORMAr Root RootAr  \
0      1   6    3     2    musotaqiyma  ADJ    مُسْتَقِيمَ  qwm    قوم   
1      2  25   31     1   m~uTah~arapN  ADJ   مُّطَهَّرَةٌ  Thr    طهر   
2      2  65   12     1    xa`si_#iyna  ADJ    خَٰسِـِٔينَ  xsA    خسا   
3      2  68   15     1        faAriDN  ADJ        فَارِضٌ  frD    فرض   
4      2  69   15     1        faAqiEN  ADJ        فَاقِعٌ  fqE    فقع   
..   ...  ..  ...   ...            ...  ...            ...  ...    ...   
330  101   7    4     1     r~aADiyapK  ADJ     رَّاضِيَةٍ  rDw    رضو   
331  101  11    2     1     HaAmiyapN[  ADJ     حَامِيَةٌۢ  Hmy    حمي   
332  104   6    3     2      muwqadapu  ADJ      مُوقَدَةُ  wqd    وقد   
333  104   9    3     1  m~umad~adapK]  ADJ  مُّمَدَّدَةٍۭ  mdd    مدد   
334  105   5    3     1    m~a>okuwlK]  ADJ    مَّأْكُولٍۭ  Akl    اكل   

           Lemma      LemmaAr Derivation  State Grop Case FormN  
0    m~usotaqiym  مُّسْتَقِيم   ACT PCPL    DEF    M  ACC     X  
1    m~uTah~arap  مُّطَهَّرَة  PASS PCPL  INDEF       NOM    II  
2         xaAsi}       خَاسِئ   ACT PCPL    DEF   MP  ACC     I  
3         faAriD       فَارِض   ACT PCPL  INDEF       NOM     I  
4         faAqiE       فَاقِع   ACT PCPL  INDEF       NOM     I  
..           ...          ...        ...    ...  ...  ...   ...  
330     raADiyap     رَاضِيَة   ACT PCPL  INDEF       GEN     I  
331     HaAmiyap     حَامِيَة   ACT PCPL  INDEF       NOM     I  
332     muwqadap     مُوقَدَة  PASS PCPL    DEF    F  NOM    IV  
333  m~umad~adap  مُّمَدَّدَة  PASS PCPL  INDEF       GEN    II  
334    m~a>okuwl    مَّأْكُول  PASS PCPL  INDEF       GEN     I  

[335 rows x 16 columns]